prov/verbs/src/windows/verbs_nd_rdma.c

```diff
@@ -0,0 +1,710 @@
+	}
+
+	struct adapter_t *adapter = NULL;
+	int i;
+	for (i = 0; i < vrb_nd_infra.adapters.count; ++i) {
+		if (!strncmp(name, vrb_nd_infra.adapters.adapter[i].name,
+			     strlen(name))) {
+			id->verbs = vrb_nd_infra.adapters.context_list[i];
+			adapter = vrb_nd_infra.adapters.adapter_list[i];
+			id->route.addr.src_addr = *addr;
+			break;
+		} else if (!strncmp(name, "127.0.0.1", strlen(name)) &&
+			   (vrb_nd_infra.adapters.adapter[i].info.AdapterFlags &
+			    ND_ADAPTER_FLAG_LOOPBACK_CONNECTIONS_SUPPORTED)) {
+			id->route.addr.src_addr = *addr;
+			break;
+		}
+	}
+	if (i == vrb_nd_infra.adapters.count) {
+		errno = ENODEV;
+		return -1;
+	}
+
+	if (adapter) {
+		struct rdma_cm_id_vrb_nd *id_nd =
+			container_of(id, struct rdma_cm_id_vrb_nd, id);
+		HRESULT hr = adapter->adapter->lpVtbl->CreateOverlappedFile(
+			adapter->adapter, &id_nd->ov_file);
+		FI_LOG(&vrb_prov, FAILED(hr) ? FI_LOG_WARN : FI_LOG_DEBUG,
+		       FI_LOG_EP_CTRL,
+		       "IND2Adapter::CreateOverlappedFile: hr=0x%08lx\n", hr);
+		if (FAILED(hr)) {
+			errno = H2F(hr);
+			return -1;
+		}
+
+		assert(id_nd->ov_file &&
+		       id_nd->ov_file != INVALID_HANDLE_VALUE);
+
+		BindIoCompletionCallback(id_nd->ov_file, vrb_nd_io_cb, 0);
+
+		hr = adapter->adapter->lpVtbl->CreateConnector(
+			adapter->adapter, &IID_IND2Connector, id_nd->ov_file,
+			(void **)&id_nd->connector);
+		FI_LOG(&vrb_prov, FAILED(hr) ? FI_LOG_WARN : FI_LOG_DEBUG,
+		       FI_LOG_EP_CTRL,
+		       "IND2Adapter::CreateConnector: hr=0x%08lx\n", hr);
+		if (FAILED(hr)) {
+			int ret = CloseHandle(id_nd->ov_file);
+			assert(ret);
+			errno = H2F(hr);
+			return -1;
+		}
+	}
+
+	return FI_SUCCESS;
+}
+
+int rdma_resolve_addr(struct rdma_cm_id *id, struct sockaddr *src_addr,
+		      struct sockaddr *dst_addr, int timeout_ms)
+{
+	trace(__func__);
+	if ((!id) || (!dst_addr)) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	if (!id->verbs) {
+		if (!src_addr) {
+			if (vrb_nd_infra.adapters.count == 0) {
+				errno = ENODEV;
+				return -1;
+			}
+			src_addr =
+				&vrb_nd_infra.adapters.adapter[0].address.addr;
+		}
+
+		if (rdma_bind_addr(id, src_addr)) {
+			return -1;
+		}
+
+		struct rdma_cm_id_vrb_nd *id_nd =
+			container_of(id, struct rdma_cm_id_vrb_nd, id);
+		HRESULT hr = id_nd->connector->lpVtbl->Bind(
+			id_nd->connector, &id->route.addr.src_addr,
+			sizeof(id->route.addr.src_addr));
+		FI_LOG(&vrb_prov, FAILED(hr) ? FI_LOG_WARN : FI_LOG_DEBUG,
+		       FI_LOG_EP_CTRL, "IND2Connector::Bind: hr=0x%08lx\n", hr);
+		if (FAILED(hr)) {
+			errno = H2F(hr);
+			return -1;
+		}
+
+		ULONG addrlen = sizeof(id->route.addr.src_addr);
+		hr = id_nd->connector->lpVtbl->GetLocalAddress(
+			id_nd->connector, &id->route.addr.src_addr, &addrlen);
+		FI_LOG(&vrb_prov, FAILED(hr) ? FI_LOG_WARN : FI_LOG_DEBUG,
+		       FI_LOG_EP_CTRL,
+		       "IND2Connector::GetLocalAddress: hr=0x%08lx\n", hr);
+		if (FAILED(hr)) {
+			errno = H2F(hr);
+			return -1;
+		}
+	}
+
+	id->route.addr.dst_addr = *dst_addr;
```

Comment:
sockaddr isn't a real address.  We should check the sa_family and copy the address based on the size sockaddr_in or sockaddr_in6 or sockaddr_ib)